src/python/pants/backend/terraform/target_gen.py

```diff
@@ -0,0 +1,70 @@
+# Copyright 2021 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+from __future__ import annotations
+
+from pants.backend.terraform.target_types import (
+    TerraformModule,
+    TerraformModules,
+    TerraformModulesSources,
+)
+from pants.build_graph.address import Address
+from pants.core.goals.tailor import group_by_dir
+from pants.engine.internals.selectors import Get
+from pants.engine.rules import collect_rules, rule
+from pants.engine.target import (
+    GeneratedTargets,
+    GenerateTargetsRequest,
+    ImmutableValue,
+    Sources,
+    SourcesPaths,
+    SourcesPathsRequest,
+    Target,
+)
+from pants.engine.unions import UnionRule
+
+
+class GenerateTerraformModuleTargetsRequest(GenerateTargetsRequest):
+    generate_from = TerraformModules
+
+
+@rule
+async def generate_terraform_module_targets(
+    request: GenerateTerraformModuleTargetsRequest,
+) -> GeneratedTargets:
+    sources_paths = await Get(
+        SourcesPaths, SourcesPathsRequest(request.generator.get(TerraformModulesSources))
+    )
+
+    dir_to_filenames = group_by_dir(sources_paths.files)
+    dirs_with_terraform_files = []
+    for dir, filenames in dir_to_filenames.items():
+        if any(filename.endswith(".tf") for filename in filenames):
+            dirs_with_terraform_files.append(dir)
+
+    def gen_target(dir: str) -> Target:
+        generated_target_fields = {}
+        for field in request.generator.field_values.values():
+            value: ImmutableValue | None
+            if isinstance(field, Sources):
+                # TODO: Should this just use a glob if all of the filenames match the default *.tf glob?
+                value = tuple(sorted(dir_to_filenames[dir]))
```

Comment:
Because this gets recalculated every time the relevant parts of file system change—and it never gets saved to disk beyond running `./pants peek prod/terraform:mods#subdir`—no, I don't think you need to use a glob. It makes sense to be explicit.

> Because this gets recalculated every time the relevant parts of file system change

Meaning, the generated target will change when a new file is added or a file is removed.

Reply:
Makes sense. Will remove the TODO.